src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -646,16 +664,30 @@ public void doStatus(@QueryParameter String token) {
             System.err.println("Checking status");
             checkToken(token);
         }
+        /**
+         * Used to run test methods on a separate thread so that code that uses {@link Stapler.getCurrentRequest}
+         * does not inadvertently interact with the request for {@link #doStep} itself.
+         */
+        private static final ExecutorService STEP_RUNNER = Executors.newSingleThreadExecutor(
+                new NamingThreadFactory(new DaemonThreadFactory(), RealJenkinsRule.class.getName() + ".STEP_RUNNER"));
         @POST
         public void doStep(StaplerRequest req, StaplerResponse rsp) throws Throwable {
             List<?> tokenAndStep = (List<?>) Init2.readSer(req.getInputStream(), Endpoint.class.getClassLoader());
             checkToken((String) tokenAndStep.get(0));
             Step s = (Step) tokenAndStep.get(1);
             Throwable err = null;
-            try (CustomJenkinsRule rule = new CustomJenkinsRule(); ACLContext ctx = ACL.as(ACL.SYSTEM)) {
-                s.run(rule);
-            } catch (Throwable t) {
-                err = t;
+            try {
+                STEP_RUNNER.submit(() -> {
+                    try (CustomJenkinsRule rule = new CustomJenkinsRule(); ACLContext ctx = ACL.as(ACL.SYSTEM)) {
+                        s.run(rule);
+                    } catch (Throwable t) {
+                        throw new RuntimeException(t);
+                    }
+                }).get();
+            } catch (ExecutionException e) {
+                err = e.getCause().getCause();
```

Comment:
```suggestion
                // Once for ExecutionException, once for RuntimeException:
                err = e.getCause().getCause();
```
May be more robust and clearer to use an `AtomicReference<Throwable>` though.

Reply:
I added a comment in [869ed42](https://github.com/jenkinsci/jenkins-test-harness/pull/422/commits/869ed42113bfc9b6f3d3a4f70e26fa1841e7fc3e). I experimented with `AtomicReference<Throwable>` but it was less clear in my opinion.